compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1999,70 +2022,106 @@ TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::C
          // decomposition failed
          // large constants must be loaded into a register first so these cases
          // are evaluated by the generic analyser call below
-         if (targetRegister == 0 && IS_32BIT_SIGNED(value)) // decomposition failed
+         if (targetRegister == 0 && IS_32BIT_SIGNED(value) && (dataType != TR::Int8)) // decomposition failed
             {
-            TR_X86OpCodes opCode;
-            if (firstChild->getReferenceCount() > 1 ||
-                firstChild->getRegister() != 0)
-               {
-               if (value >= -128 && value <= 127)
-                  {
-                  opCode = IMULRegRegImms(nodeIs64Bit);
-                  }
+            TR_X86OpCodes opCode = BADIA32Op;
+           
+            // Genererate Opcodes 
+            switch(dataType)
+            {
+            case TR::Int16:
+               if(isMemRef)
+                  opCode = (value >= -128 && value <= 127) ? IMUL2RegMemImms : IMUL2RegMemImm2;
                else
+                  opCode = (value >= -128 && value <= 127) ? IMUL2RegRegImms : IMUL2RegRegImm2;
+               break;
+            case TR::Int32:
+            case TR::Int64:
+               if(value >= TR::getMinSigned<TR::Int32>() && value <= TR::getMaxSigned<TR::Int32>())
                   {
-                  opCode = IMULRegRegImm4(nodeIs64Bit);
+                  if(isMemRef)
```

Comment:
There is a bug with this because when `firstChild->getReferenceCount() > 1 ` even with `isMemRef == true` we would use the `generateRegRegImmInstruction` version. With the current code we can end up passing `IMULRegMem***` opcode to `generateRegRegImmInstruction`.  I would suggest keeping the old code:
```
if (firstChild->getOpCode().isMemoryReference())
{
   if (value >= -128 && value <= 127)
   {
      opCode = IMULRegMemImms(nodeIs64Bit); define a new query IMULRegMemImms that takes Int16, Int32, Int64 instead of just a boolean and return corresponding opcode.
   }
   else
   {
      opCode = IMULRegMemImm4(nodeIs64Bit); same for this one
   }
}
```